Docs: Integration

sentientwaffle opened the pull request:
Document integration guidelines:
- Consistency guarantees from the application and client session's point of view.
- Advice about batching.
- Data modelling — ideas to take advantage of TigerBeetle's invariants and indexes.

sentientwaffle committed:
Docs: Integration

Document integration guidelines:
- Consistency guarantees from the application and client session's point of view.
- Advice about batching.
- Data modelling — ideas to take advantage of TigerBeetle's invariants and indexes.

eatonphil commented on docs/usage/integration.md:
```diff
@@ -0,0 +1,413 @@
+Strict consistency guarantees at the database level simplify application logic and error handling
+farther up the stack.
+
+
```
No need for more than two newlines here (there are 4?)

sentientwaffle replied:
Yeah, I do that deliberately, it helps me keep the sections clearly separate in my head as I jump around & edit. It won't show up in the rendered markdown.

eatonphil commented on docs/usage/integration.md:
```diff
@@ -0,0 +1,413 @@
+
+TigerBeetle is designed to guard against bugs not only in its [own code](../TIGER_STYLE.md), but
+at the boundaries, in the application code which interfaces with TigerBeetle.
+This is exhibited by the client's API design, which may be [surprising](#retries) when contrasted
```
Instead of a link without a description maybe change this to `which may be surprising (see [#retries](#retries))`

eatonphil commented on docs/usage/integration.md:
```diff
@@ -0,0 +1,413 @@
+TigerBeetle is designed to guard against bugs not only in its [own code](../TIGER_STYLE.md), but
+at the boundaries, in the application code which interfaces with TigerBeetle.
+This is exhibited by the client's API design, which may be [surprising](#retries) when contrasted
+with a more conventional database.
```
Are retries really so different from conventional databases?

sentientwaffle replied:
The client doesn't have a retry limit — it will never bubble a network error up to the application, it will just keep trying.

eatonphil commented on docs/usage/integration.md:
```diff
@@ -0,0 +1,413 @@
+This is exhibited by the client's API design, which may be [surprising](#retries) when contrasted
+with a more conventional database.
+
+Strict consistency guarantees at the database level simplify application logic and error handling
```
Since there are a few phrases in there, changing to `Strict consistency guarantees (at the database level) simplify` might help the reader parse

eatonphil commented on docs/usage/integration.md:
```diff
@@ -0,0 +1,413 @@
+[client session](#client-sessions).
+
+But consistency models are notoriously arcane — what specifically does this mean for applications
+in practice?
```
Woohoo! Clarifying arcaneness

jorangreef replied:
Serializability is roughly "as if transactions are executed serially, as if nothing else is executing concurrently, so that a transaction gets a view of all the objects in the world, frozen in time, with nothing concurrently modifying them" (this doesn't prevent the database from executing transactions concurrently if they don't overlap, also doesn't guarantee order in which transactions are executed).

Linearizability is roughly "the execution w.r.t. an object happens between operation request and response (and doesn't jump around in time before or after that)".

These two are at the top of the pyramid of all the different types of consistency models so they're common and not arcane. It's beneath this where things start to get hairy and have different names.

Here's a great explanation from Peter Bailis: http://www.bailis.org/blog/linearizability-versus-serializability/

jorangreef replied:
Would be great to bring his definitions in with a link back.

sentientwaffle replied:
Imo serializability is still arcane. That is why this section avoids explaining them at a high level it all — it gives specific behaviors that the application developer can expect from TB.

I'm happy to link to that explanation, but I don't want to distract from the list of behaviors.

eatonphil replied:
> so they're common and not arcane

Yeah I agree with DJ that these terms and meanings are still not well known. Maybe "arcane" is too harsh. I've read DDIA and read about databases how many times and I still need the refresher each time I look at it.

sentientwaffle replied:
My original draft said "obtuse" instead of "arcane" :wink:.

eatonphil replied:
obtuse is also harsh :D 

Instead of:

> But consistency models are notoriously arcane — what specifically does this mean for applications
in practice?

Maybe

> But consistency models are easy to misunderstand or forget about — what specifically does this mean for applications in practice?

"obtuse" and "arcane" can sound like we're putting down the researchers or database developers. Rather, it's just about being gracious to the audience by way of a reminder

eatonphil replied:
"forget about" isn't quite right either, but something like that

jorangreef replied:
That's great.

sentientwaffle replied:
:thinking: You're right, this sentence needs to be reworked..

"easy to misunderstand" is a little too close to "hard to understand" — which is accurate from my perspective, but is still too close to the obtuse/arcane.

What do you think of:

> But consistency models are easy to misunderstand and hard to apply.
> What specific guarantees does TigerBeetle provide to applications?

I don't love the "unfamiliar systems" part...

jorangreef replied:
Somewhere we should just have a nice section on guarantees w.r.t. consistency, and clearly state "strict serializability" and then explain how this is comprised of "serializability" and "linearizability" and what those mean. I like how Kyle Kingsbury always looks for this when he does engagements—that the database has a clear discussion of guarantees.

It is true also that after these big three, things get into the weeds pretty quickly. For example, even in the DistSys community, not everyone agrees even on the taxonomy.

sentientwaffle replied:
Is TigerBeetle's documentation really the place to explain core dist sys concepts though? I agree we should specify what we provide (strict serializability), but if the reader isn't familar with those, it seems like pointing them in the direction of e.g. Bailis' guide is fine.

eatonphil replied:
If the original sentence only said "But consistency models can seem arcane" I'd be fine with everything else as is and also with the Bailis link.

jorangreef replied:
> Is TigerBeetle's documentation really the place to explain core dist sys concepts though?

We don't have to go deep. A quick intro to all 3 in the right section in the right doc would be worthwhile. Serializability and linearizability together help to explain strict serializability.

sentientwaffle replied:
:+1: 

What do you think of putting the linearizability/serializability explanation in a separate (future — I'll write myself a TODO) document that we can link to, though?

It could be part of an "internals of TigerBeetle's consensus" document which gives an overview of our use (and modifications to) VSR, how it relates to the superblock, and client table.

That way `integration.md` focuses on practical "how does a client correctly use TigerBeetle" — to stay as concrete as possible.

jorangreef replied:
Perfect. We should just somewhere have a very visible simple guarantees area, where we highlight "strict serializability" as the consistency level, since that's something that should be on "the outside of the box" so that people know the big ideas. i.e. That they're not getting eventual consistency.

eatonphil commented on docs/usage/integration.md:
```diff
@@ -0,0 +1,413 @@
+in practice?
+
+- [**Client sessions**](#client-sessions)
+  - A client session may have at most one in-flight request.
```
I don't want to get you bogged down in the weeds. But is `in-flight` a generally understood idea?

sentientwaffle replied:
:thinking: I thought it was, but I don't know for sure. Do you think "pending" would be more clear?

eatonphil replied:
Is `may have at most one request at a time` equivalent?

sentientwaffle replied:
I think that is a little less accurate, clients do have multiple requests, they just queue them up locally until the in-flight/pending one finishes.

eatonphil replied:
ok!

eatonphil commented on docs/usage/integration.md:
```diff
@@ -0,0 +1,413 @@
+  - A client session reads its own writes.
+  - A client session observes writes in the order that they occur on the cluster.
+  - A client session observes [`debits_posted`](../reference/accounts.md#debits_posted) and
+    [`credits_posted`](../reference/accounts.md#credits_posted) as monotonically increasing.
```
Maybe one more followup sentence to clarify further? `as monotonically increasing. Within a client session, you will never see credits_posted or debits_posted decrease.` Though you are trying to avoid `you`...

sentientwaffle replied:
:+1: Good clarification, "monotonically increasing" is still a little obtuse. How about

> That is, a client session will never see `credits_posted` or `debits_posted` decrease.

eatonphil replied:
Yep that's good

eatonphil commented on docs/usage/integration.md:
```diff
@@ -0,0 +1,413 @@
+  - Requests retried by a different client (same request body, different session) may receive
+    [different replies](#consistency-with-foreign-databases).
+  - Operations within a request are executed in sequential order.
+  - Operations within a request do not interleave with other requests.
```
This is already given as not possible because only one request can be inflight right? Unless you mean with respect to other concurrent client sessions?

sentientwaffle replied:
> Unless you mean with respect to other concurrent client sessions?

Yeah, that is what I meant.

How about:

>   - Operations within a request do not interleave with other requests' operations.

eatonphil replied:
Why not mention other client requests? Why keep that part implicit?

sentientwaffle replied:
Because I didn't think of that :sweat_smile:.

How about:

>   - Operations within a request do not interleave with operations from other clients' requests.

My only worry is that this might sound like they could interleave with operations from their own requests.

sentientwaffle replied:
Also, I am going to add a new bullet in this same vein:

>   - A client session can consider a request executed when it receives a reply for the request.

eatonphil replied:
> Operations within a request do not interleave with operations from other clients' requests.
> My only worry is that this might sound like they could interleave with operations from their own requests.

Yeah I'm now confused what the real behavior is.

eatonphil commented on docs/usage/integration.md:
```diff
@@ -0,0 +1,413 @@
+  - [Transfers](../reference/transfers.md) are immutable — once created, they are never modified.
+  - Transfer [timeouts](../reference/transfers.md#timeout) are deterministic, driven by the
+    cluster's timestamp.
+  - Object timestamps are cluster-unique. For objects `A` and `B`, `A.timestamp ≠ B.timestamp`.
```
`For all objects`?

eatonphil commented on docs/usage/integration.md:
```diff
@@ -0,0 +1,413 @@
+  - Transfer [timeouts](../reference/transfers.md#timeout) are deterministic, driven by the
+    cluster's timestamp.
+  - Object timestamps are cluster-unique. For objects `A` and `B`, `A.timestamp ≠ B.timestamp`.
+  - Object timestamps are strictly increasing For objects `A` and `B`, if
```
`For all objects`?

eatonphil commented on docs/usage/integration.md:
```diff
@@ -0,0 +1,413 @@
+
+### Retries
+
+A client will retry a request until either:
```
"client" can be confusing on its own. Maybe we should define at the top when we say "client" we mean client library? Unless you actually mean user code? I'm not actually sure which you mean here.

sentientwaffle replied:
I'll change that to "client session". I try to refer to "client sessions" directly (instead of "clients") because "client" is so overloaded in databases.

eatonphil replied:
I'm still confused though. When you say "will" are you saying that TigerBeetle client libraries do this for you or that the application developer will have to put retries in place?

sentientwaffle replied:
TigerBeetle client libraries do this automatically.

How about:

> A client session will automatically retry a request until either:

eatonphil commented on docs/usage/integration.md:
```diff
@@ -0,0 +1,413 @@
+
+- the TigerBeetle client will never time out
+- the TigerBeetle client has no retry limits
+- the TigerBeetle does not surface network errors
```
No need for `the` in this last bullet?

sentientwaffle replied:
I'll add "client" there instead.

eatonphil commented on docs/usage/integration.md:
```diff
@@ -0,0 +1,413 @@
+- the TigerBeetle does not surface network errors
+
+With TigerBeetle's strict consistency model, surfacing these errors at the client/application level
+would be misleading, because they would imply that a request did not execute, when that is not known:
```
Long sentence could be broken (if you want): `would be misleading. They would imply`

eatonphil replied:
And although strictos don't like it, it is actually defensible to start a sentence with "And" or "Because" :D

eatonphil replied:
ehh nevermind that would make it not a sentence. Yeah if you break it up you could just drop `Because`

sentientwaffle replied:
Breaking it up is a good suggestion, I'll do:

> With TigerBeetle's strict consistency model, surfacing these errors at the client/application level
> would be misleading. An error would imply that a request did not execute, when that is not known:

eatonphil replied:
looks good!

sentientwaffle committed:
Docs: Review fixes; split out Data Modeling

batiati commented:
I ran out of adjectives .... Bravo, Bravissimo 👏👏👏

sentientwaffle commented:
Thanks @batiati! This document was heavily inspired by our discussion.

sentientwaffle committed:
Docs: review update

jamii commented on docs/usage/integration.md:
```diff
@@ -0,0 +1,314 @@
+  - [Transfers](../reference/transfers.md) are immutable — once created, they are never modified.
+  - Transfer [timeouts](../reference/transfers.md#timeout) are deterministic, driven by the
+    cluster's timestamp.
+  - Object timestamps are cluster-unique. For all objects `A` and `B`, `A.timestamp ≠ B.timestamp`.
```
```suggestion
  - Within a cluster, object timestamps are unique. For all objects `A` and `B` stored in the same cluster, `A.timestamp ≠ B.timestamp`.
```

sentientwaffle replied:
Ooh, good catch. I'll reword the next line likewise.

jamii commented on docs/usage/integration.md:
```diff
@@ -0,0 +1,314 @@
+  - Transfer [timeouts](../reference/transfers.md#timeout) are deterministic, driven by the
+    cluster's timestamp.
+  - Object timestamps are cluster-unique. For all objects `A` and `B`, `A.timestamp ≠ B.timestamp`.
+  - Object timestamps are strictly increasing For all objects `A` and `B`, if
```
```suggestion
  - Object timestamps are strictly increasing. For all objects `A` and `B`, if
```

jamii commented on docs/usage/integration.md:
```diff
@@ -0,0 +1,314 @@
+  - Object timestamps are cluster-unique. For all objects `A` and `B`, `A.timestamp ≠ B.timestamp`.
+  - Object timestamps are strictly increasing For all objects `A` and `B`, if
+    `A.timestamp < B.timestamp`, then `A` was committed earlier than `B`.
+
```
Maybe add something here about consistency guarantees between sessions. Eg if my client crashes and I restart on the same machine, am I guaranteed to see writes from the previous session. If not, how do I deal with that from an application point of view.

jamii commented on docs/usage/integration.md:
```diff
@@ -0,0 +1,314 @@
+
+## Requests
+
+A _request_ is a [batch](#batching-operations) of one or more operations sent to the cluster in a
```
Should we s/operation/events to keep docs consistent with the code?

sentientwaffle replied:
:thinking: we do use Event in places, but not very widely right now.

@jorangreef do you have a preference for the terminology? i.e. is an "event"/"operation event" the request-side of an operation? We could call it an "operation request", but that might be confusing. ~But then again we use "reply" for both the individual operation reply, and the batch of replies.~ "Result" is for individual replies, "Reply" is for the batch.

sentientwaffle committed:
Docs: Clarify session restart consistency

sentientwaffle committed:
Docs: Accounting Operations

Rename `docs/operations` to `docs/deployment` to avoid overloading "operations".
Other alternatives: "administration".

This links to the "Asset Exchange" explanation which isn't finished/committed yet.

sentientwaffle committed:
Docs: Fractional Amounts

sentientwaffle committed:
Docs: Asset Exchange guide

sentientwaffle committed:
Docs: Manifest, ManifestLevel, ManifestLog

sentientwaffle committed:
Docs: Integration: Distinguish operations/events

sentientwaffle committed:
Docs: Transfers: Two-phase transfers

- Add a "modes" section of the reference document. This allows information on two-phase transfers (and the contrast with single-phase transfers) to be more linear & contiguous, which is easier to understand.
- Elaborate on two-phase transfers, with more explanation & examples.
- Fix "constraints" of fields which may be zero for post/void transfers.
- Fix links with "#flags." — "." isn't valid in CSS ids, so it isn't included in URL fragments.
- Link to `operations/*.md`.

sentientwaffle committed:
Docs: Add and fix links

URL fragments can't/don't include a ".".

sentientwaffle committed:
Docs: Data Modeling: Debits/Credits explanation

Also:
- "id" still needs more work, since right now it describes advanced usage without exploring (or explaining) the common case.
- Add links to new docs and Slack.

sentientwaffle committed:
Docs: ECC memory recommendation

sentientwaffle committed:
Docs: Document UUIDv4 recommendation

sentientwaffle committed:
Docs: typo

eatonphil commented on README.md:
```diff
@@ -19,6 +19,27 @@ structures.
 
+## Documentation
+
+- Setup
```
I don't really want to keep this updated manually as opposed to just linking people to docs.tigerbeetle.com?

sentientwaffle committed:
Docs: UUIDv4 → "Random Identifier"

Don't recommend UUIDv4, just a random id (via CSPRNG) — UUIDv4 has a few fixed bits.

Also:
- Fix links that Docusaurus found.
- Trim the documentation listed in the README.
- In README, link directly to docs.tigerbeetle.com instead of the markdown files (better navigation).

eatonphil commented on README.md:
```diff
@@ -23,6 +23,17 @@ structures.
 
+## Documentation
+
+- Deployment
```
Maybe add a sentence:

Check out docs.tigerbeetle.com. Here are a few key pages you might be interested in.

?

eatonphil commented on README.md:
```diff
@@ -197,7 +208,10 @@ accounted for!
 For further reading:
 
-* [Running a 3-node cluster locally with docker-compose](./docs/setup/with-docker-compose.md)
+* [Creating accounts and transfers in the Node CLI](https://docs.tigerbeetle.com/usage/node-cli)
```
The node-cli doc is a complete duplicate of what is in the readme fwiw

sentientwaffle replied:
Ok, I'll remove this link.

eatonphil commented on docs/reference/operations/create_accounts.md:
```diff
@@ -0,0 +1,112 @@
+# `create_accounts`
+
+Create one or more [`Account`s](../accounts.md).
```
very minor nit: I'd put the `s` outside the link

eatonphil commented on docs/reference/operations/create_accounts.md:
```diff
@@ -0,0 +1,112 @@
+error is applicable, only the one listed first is returned.
+
+### `ok`
+The account was successfully created, but did not previously exist.
```
`but` is a little confusing. Maybe `created; it did not previously exist`?

eatonphil commented on docs/reference/operations/create_accounts.md:
```diff
@@ -0,0 +1,112 @@
+
+The following flags are mutually exclusive:
+- [`Account.flags.debits_must_not_exceed_credits`](../accounts.md#flagsdebits_must_not_exceed_credits)
+- [`Account.flags.credits_must_not_exceed_debits`](../accounts.md#flagscredits_must_not_exceed_debits)
```
Dumb question: why are these mutually exclusive? It's not possible you'd want to make sure an account is completely balanced? Maybe that's nonsense. Just wondering

sentientwaffle replied:
If both `exceed` flags were set, you would never be able to change its balance.

eatonphil replied:
You could, as part of a (database-style) transaction (which I guess are linked events), enforce that an account always gets transferred some amount when it transfers some amount out.

sentientwaffle replied:
Well, not with how the state machine works right now :wink:.
In particular, that would change the state machine semantics, which currently are simple because each transfer is executed serially, and its constraints are valid at the end of every transfer. For "equal-in-and-out" to work, transfers must logically execute simultaneously, which is harder to reason about.

eatonphil commented on docs/reference/operations/create_accounts.md:
```diff
@@ -0,0 +1,112 @@
+- `debits_pending`
+- `debits_posted`
+- `credits_pending`
+- `credits_posted`
```
Isn't it only `id` that must be identical? You don't need all other fields to be equal for us to register it as existing right?

sentientwaffle replied:
There are other variants of the `exists` error: `exists_with_different_ledger`, `exists_with_different_code`, etc. So if `ledger`/`code` differ, those respective errors are returned instead of `exists`.

`{debits,credits}_{pending,posted}` are a special case because clients shouldn't be creating accounts with nonzero balances anyway — that is a separate (higher-priority) error, because it implements a bug in the application or client code.

eatonphil replied:
I'm still confused though, do we try to say that things are equal if fields other than id are equal? Isn't id the only important thing in discovering existence/not?

sentientwaffle replied:
For all `exists_...` variants the `id` matches — that's a prerequisite.
We return a more specific error when certain fields (e.g. `ledger`) are mismatched, because that probably indicates a programmer error.

Just plain `exists` doesn't (necessarily) indicate a programmer error though — it is normal when the application is recovering from a crash and replaying operations because it isn't sure where it left off.

I'll rephrase this a bit to make that more clear.

eatonphil commented on docs/reference/operations/create_accounts.md:
```diff
@@ -0,0 +1,112 @@
+
+To correctly recover from application crashes
+[many applications](../../usage/integration.md#conistency-with-foreign-databases) should handle
+`exists` exactly as [`ok`](#ok).
```
Dumb question: why does `exists` exist then if the correct way to handle it is only ever to treat it as ok?

sentientwaffle replied:
Not a dumb question at all — actually an extremely tricky one!

- Treating `exists` as `ok` is correct in this particular context. (i.e. when coordinating with a foreign database as the "conistency-with-foreign-databases" describes).
- But there are other situations where `exists` should be treated differently. I just haven't documented them yet (they are quite niche — this section describes how integration will work most of the time).
- Part of it is also TB's philosophy of returning the most detailed result code it can. (e.g. rather than just `error`, we might send `reserved_flag`).

eatonphil commented on docs/reference/operations/create_transfers.md:
```diff
@@ -0,0 +1,309 @@
+## Result
+
+Results are listed in this section in order of descending precedence — that is, if more than one
+error is applicable, only the one listed first is returned.
```
This is just a little bit unclear that you're switching between just listing them and talking about a specific instance of a call. Maybe

> Results are listed in this section in order of descending precedence — that is, if more than one error is applicable for a transfer being created, only the one listed first error is returned.

?

eatonphil commented on docs/reference/operations/create_transfers.md:
```diff
@@ -0,0 +1,309 @@
+[`Transfer.credit_account_id`](../transfers.md#credit_account_id) must have an identical
+[`ledger`](../accounts.md#ledger).
+
+(TODO: Link to explanation of implementing currency exchange with liquidity accounts).
```
Is this TODO intended to be merged?

sentientwaffle replied:
Oops; I'll implement that now. Thanks.

eatonphil commented on docs/reference/operations/create_transfers.md:
```diff
@@ -0,0 +1,309 @@
+differ from the [`Transfer.ledger`](../transfers.md#ledger).
+
+### `exists_with_different_flags`
+The transfer already exists, but with different [`flags`](../transfers.md#flags).
```
Clarifying here as well, by transfer already exists we are just talking about the transfer's id right?

sentientwaffle replied:
Yes, I'll clarify that.

eatonphil commented on docs/reference/operations/create_transfers.md:
```diff
@@ -0,0 +1,309 @@
+
+### `overflows_debits_pending`
+The transfer was not created.
+`debit_account.debits_pending + transfer.amount` would overflow a 64-bit unsigned integer.
```
Not that we necessarily need to cover it in this PR but what is supposed to happen in business logic in this case?

sentientwaffle replied:
u64's are very big... if they hit this limit it is probably an application bug, or their data model was designed wrong (i.e. they should have used a different asset scale). Maybe in some cryptocurrency use cases, where there are extremely high-precision currencies, you may have to have multiple `ledger`s for the same currency, where they are different scales. (e.g. store "cents" and "dollars" in separate accounts, with different scales). That would be extra work for the application, but it is an extremely niche use case.

eatonphil commented on docs/reference/operations/index.md:
```diff
@@ -0,0 +1,30 @@
+# Operations
```
Could you add a _category_.json file and position this last (position 3) among the three files/folders in Reference? Right now it splits Accounts and Transfers in the docs site

eatonphil commented on docs/reference/operations/index.md:
```diff
@@ -0,0 +1,30 @@
+| ------------------ | -------------------------------------------- | ------------------------------------------------------ |
+| `create_accounts`  | [`Account`](./create_accounts.md#Event)      | [`CreateAccountResult`](./create_accounts.md#Result)   |
+| `create_transfers` | [`Transfer`](./create_transfers.md#Event)    | [`CreateTransferResult`](./create_transfers.md#Result) |
+| `lookup_accounts`  | [`Account.id`](./lookup_accounts.md#Event)   | [`Account`](./lookup_accounts.md#Result) or nothing    |
```
Maybe change Result of the last two to `List of ['Account'](...)`? (inner ticks replaced with apostrophe because I can't escape inner ticks)

sentientwaffle replied:
A `Result` matches 1:1 with an Event. This is a table of Result type, so both columns are singular, and would be batched.

sentientwaffle replied:
It is a question of terminology. We say that a Result is singular, and a batch of Results is called a Reply.

eatonphil replied:
gotcha

eatonphil replied:
Then maybe change the column headers to say `Event (Batched)` and `Result (Batched)` or something like that since the operation create_accounts/whatnot always deals with lists.

sentientwaffle replied:
I'll change the previous sentence to:

> Client implementations provide an API for sending batched events and decoding the corresponding batched results.

I want to leave the table in the singular (unbatched); it will match better with the API once clients implement batching internally.

eatonphil commented on docs/reference/operations/lookup_accounts.md:
```diff
@@ -0,0 +1,12 @@
+
+An [`id`](../accounts.md#id) belonging to a [`Account`](../accounts.md).
+
+## Result
```
This is a little confusing in that the result is always a list

sentientwaffle replied:
I think it will make more sense once client implementations handle batching — then from the application's point of view, event/result will be singular, and batching more of an internal implementation detail.

eatonphil commented on docs/reference/operations/lookup_transfers.md:
```diff
@@ -0,0 +1,12 @@
+
+An [`id`](../transfers.md#id) belonging to a [`Transfer`](../transfers.md).
+
+## Result
```
Same note about being a list

eatonphil commented on docs/reference/transfers.md:
```diff
@@ -3,14 +3,150 @@
+| `flags.void_pending_transfer` | false        | false    | false        | true         |
+| `amount`                      | required     | required | optional     | optional     |
+
 TigerBeetle uses the same data structures internally and
 externally. This means that sometimes you need to set temporary values
 for fields that TigerBeetle, not you (the user), are responsible.
```
This section has kind of gotten lost now. I'd suggest moving it to above the updates section like it was before or removing it

sentientwaffle replied:
I was thinking it kind of fits here actually, because I just listed out which fields are required (or required to be omitted, e.g. `none`).

eatonphil commented on docs/reference/transfers.md:
```diff
@@ -3,14 +3,150 @@
+### Pending Transfer
+
+A pending transfer (denoted by [`flags.pending`](#flagspending) followed by a post-pending transfer,
+void-pending transfer, or a timeout is called a "two-phase transfer".
```
You have a section dedicated to Single-phase transfer which makes sense but then the two-phase transfer doesn't really have a section which feels asymmetric. On the other hand I get that these sections just refer to the type of the transfer.

sentientwaffle replied:
This is a good point. I was thinking of it in terms of the different transfer types — one section each. But I agree; I'll group those within a "Two-Phase Transfer" section instead.

eatonphil commented on docs/recipes/asset-exchange.md:
```diff
@@ -0,0 +1,52 @@
+# Asset Exchange
+
+Distinct [`ledger`](../reference/accounts.md#ledger) values typically denote distinct assets
```
Should we explain what an asset exchange is first before diving into how it's implemented in TigerBeetle?

I don't really know what it is, though you sort of give examples later on. Maybe put the examples first?

eatonphil commented on docs/recipes/fractional-amounts.md:
```diff
@@ -0,0 +1,34 @@
+# Fractional Amounts
```
Nice recipe!

eatonphil commented on docs/usage/data-modeling.md:
```diff
@@ -0,0 +1,150 @@
+data model. Which (if any) of these techniques are suitable is highly application-specific.
+
+When possible, encoding application invariants directly in TigerBeetle rather than implementing
+them in the application itself (or with a foreign database) minimizes round-trips and coordination.
```
"encoding application invariants" and "minimizes round-trips and coordination" are far away from eachother which makes it harder to follow the sentence IMO

You could rewrite to:

When possible, round trips and coordination can be minimized by encoding ...

eatonphil commented on docs/usage/data-modeling.md:
```diff
@@ -0,0 +1,150 @@
+In double-entry accounting, an account balance is either computed as either `debits - credits` or
+`credits - debits`, depending on the type of account. It is up to the application to compute
+the balance — TigerBeetle tracks the cumulative posted debits and cumulative posted credits,
+not their difference.
```
For a second this makes it sound like the application must sum up all transfers themselves.

Maybe instead:

> TigerBeetle tracks the cumulative posted debits and cumulative posted credits, not their difference. If the application needs the difference, it can be easily computed from the cumulative debits and credits tracked on the account.

Or something?

sentientwaffle replied:
I changed it to:

> TigerBeetle tracks each account's cumulative posted debits and cumulative posted credits.
In double-entry accounting, an account balance is the difference between the two — computed as
either `debits - credits` or `credits - debits` depending on the type of account. It is up to the
application to compute the balance from the cumulative debits/credits.

eatonphil replied:
`compute the balance` still sounds a little scary but it's clearer!

eatonphil commented on docs/usage/data-modeling.md:
```diff
@@ -0,0 +1,150 @@
+
+  - `balance = debits - credits` for accounts representing the database operator's assets.
+    - Known as a "debit balance".
+    - Account limits use
```
I don't exactly get what you mean by this. Are you saying they should use this flag or that they must or something else?

sentientwaffle replied:
I'll elaborate to:

>  - Accounts with limits use [`flags.credits_must_not_exceed_debits`](../reference/accounts.md#flagscredits_must_not_exceed_debits).

Not all accounts will need limits. This just specifies (for convenience) which flag matches up to which balance calculation.

eatonphil commented on docs/usage/data-modeling.md:
```diff
@@ -0,0 +1,150 @@
+  another database.
+- Set `user_data` to a group identifier for objects that will be queried together.
+- Set `user_data` to a transfer or account `id`.
+  (TODO: Can we use this for join queries via the query API, or must the application implement them?)
```
TODO intended for merge?

sentientwaffle replied:
Yeah, I left that one in on purpose. Maybe a little tacky to leave a TODO in the docs, but it is an open question, and I will want to make sure these docs get updated when we figure it out.

eatonphil commented on docs/usage/data-modeling.md:
```diff
@@ -0,0 +1,150 @@
+- An account and a transfer may share the same `id` — they belong to different "namespaces".
+- Avoid requiring a central oracle to generate each unique `id` (e.g. an auto-increment field in SQL).
+  A central oracle may become a performance bottleneck when creating accounts/transfers.
+- `id` is mostly accessed by point queries, but it is indexed for efficient range queries as well.
```
If ids are random the bit about efficient range queries doesn't make a ton of sense but we aren't requiring random ids I guess

sentientwaffle replied:
Yeah, the range query part is about non-random-ids.

eatonphil commented on docs/usage/data-modeling.md:
```diff
@@ -0,0 +1,150 @@
+- Random identifiers do not require a central oracle.
+- Only point queries are useful for fetching randomly-generated identifiers.
+
+To maximize id entropy, prefer a cryptographically-secure PRNG.
```
Commenting here as a reminder of the slack discussion

eatonphil commented on docs/usage/data-modeling.md:
```diff
@@ -0,0 +1,150 @@
+
+Set `id` to a "foreign key" — that is, reuse an identifier of a corresponding object from another
+database. For example, if every user (within the application's database) has a single account, then
+the identifier within the foreign database can be used as the `Account.id` within TigerBeetle.
```
Should we include a warning about bigserial account ids and transfer ids or is that too obscure?

sentientwaffle replied:
I'm not familiar with bigserial ids, but if it is database-specific it probably doesn't belong here. (I am curious what you mean though).

eatonphil replied:
ah sorry, autoincrement : mysql, bigserial: postgres

sentientwaffle replied:
Ah, gotcha. I don't think we need to mention it. If they are already using that in their 'foreign database' then it is actually ok, it just isn't what we recommend when starting from scratch.

eatonphil replied:
well it won't be ok if we intend to collapse namespaces. but maybe we are still sticking to strategic silence in docs here for the moment while we decide :D

sentientwaffle replied:
Oh, I forgot about that. I'll leave it for now & change it if we combine the namespaces.

eatonphil commented on docs/usage/data-modeling.md:
```diff
@@ -0,0 +1,150 @@
+coordination with the foreign database for idempotent
+[application crash recovery](./integration.md#consistency-with-foreign-databases).
+
+#### Logically Grouped Objects
```
Maybe we should add `(Advanced)` to the headers of these that are not random ids? Just so that we all the more encourage people to use random ids

sentientwaffle replied:
:thinking: I'd rather not put it in the header because it pollutes the URL fragment. I think it is ok; there are 2 different places that say "Randomly-generated identifiers are recommended for most applications.".

eatonphil replied:
It's just that by being equal weight headers they otherwise look like equally suggested

sentientwaffle replied:
How about two different "Examples" sections; "Examples" and "Examples (Advanced)"?

eatonphil replied:
Sure :+1:

sentientwaffle committed:
Docs: Code (doc?) review

Rename "asset exchange" to "currency exchange" because it is a howto, and currency exchange is the most common use case.

eatonphil reviewed (approved):
Awesome work DJ
